Json pretty print

wlan0 opened the pull request:
This PR stacks on top of my other PR for adding `admin monitor` command. This one adds jq style JSON pretty printing. 

The library under pkg/json is a fork of golang’s official encoding/json relicensed to apache v2 (verified this with @abperiasamy) 

Forking allows me to easily make this a drop in replacement of “encoding/json”. The specifications of this library are (matching the style of jq):
- Bold struct keys in blue
- Strings in green
- Numbers in default color
- Indenting and spacing matches exact jq specifications 

If everything looks good, i’ll update other commands in mc to start using this package. Here is the output of `admin monitor` using this json package

![2e07de82-88b5-40e0-904f-aace242c4e84](https://user-images.githubusercontent.com/1050521/51810438-b45f7080-225c-11e9-97c6-015b34506def.png)

Here is a comparable jq output

![ebf62458-f964-435d-a1a7-4d3c03b7dff0](https://user-images.githubusercontent.com/1050521/51810474-dfe25b00-225c-11e9-95c7-97c92a65874c.png)

harshavardhana commented:
> The library under pkg/json is a fork of golang’s official encoding/json relicensed to apache v2 (verified this with @abperiasamy)

Why are we forking? @wlan0

harshavardhana commented:
If you want to construct JSON yourself you are better off using packages such as https://github.com/tidwall/sjson, copying the entire encoding/json would be too much of a burden for us. it has a lot of historical code which is not needed and it is a purely a reflection based implementation - a better drop in replacement is https://github.com/json-iterator/go

wlan0 commented:
The relevant changes for this PR are here: https://github.com/minio/mc/pull/2661/commits/9f2399a26a441dd786d120471111c96ea17d7077

@harshavardhana As we discussed offline, this is the best solution given that there are no other drop-in replacement options for colored json output in terms of technique and effort. This solution is clean and adds only the smallest amount of the code on top of existing code. 

1. I've disabled `Unmarshal()`
2. Added a comment saying that this is only for printable output
3. renamed package to colorjson

In case anyone is wondering, We cannot put in colored objects due to the fact the existing `encoding/json` escapes color codes.

harshavardhana commented on cmd/admin-monitor.go:
```diff
@@ -0,0 +1,202 @@
+/*
+ * Minio Client (C) 2016, 2017 Minio, Inc.
```
Change this to 2019 copyright.

wlan0 replied:
ok

harshavardhana commented on pkg/colorjson/colors.go:
```diff
@@ -0,0 +1,16 @@
+package colorjson
```
Can you add the license header?

wlan0 replied:
Good catch

harshavardhana commented on pkg/colorjson/decode.go:
```diff
@@ -0,0 +1,1331 @@
+			b[w] = c
+			r++
+			w++
+
```
Perhaps this file also has to go? since we only need Marshal?

wlan0 replied:
This file has dependencies in the other ones. I've disabled the Unmarshal method, which is the entrypoint for all sorts of unmarshalling.

krisis commented on cmd/admin-monitor.go:
```diff
@@ -0,0 +1,202 @@
+  {{end}}
+EXAMPLES:
+  1. Get server cpu, mem, net and disk statistics of the 'play' server.
+       $ {{.HelpName}} play/
```
1. I am able to run `mc admin monitor play extraArg`. This should exit with invalid argument error.
```
➜  mc git:(pr2661) ✗ mc admin monitor play extraArg
●  play.minio.io:9000    

   CPU        min        avg      max
   current    14.63%      14.63%    14.64%
   historic   0.03%      2.25%    462.47%

   MEM        usage
   current    467MiB
   historic   439MiB

```

2.Alignment is off for `avg` and `max` columns.

wlan0 replied:
Good catch on #1. Will fix both issues

krisis commented on cmd/admin-monitor.go:
```diff
@@ -0,0 +1,202 @@
+// checkAdminMonitorSyntax - validate all the passed arguments
+func checkAdminMonitorSyntax(ctx *cli.Context) {
+	if len(ctx.Args()) == 0 || len(ctx.Args()) > 2 {
+		exit := 1
```
We have a global const for exit status to indicate failure, namely `globalErrorExitStatus`.

wlan0 replied:
sounds good

krisis commented on pkg/colorjson/colors.go:
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2011 The Go Authors. All rights reserved
```
Should we update the copyright to 2019?

krisis replied:
OK

krisis commented on cmd/admin-monitor.go:
```diff
@@ -0,0 +1,202 @@
+	cpuMonitors, e := client.ServerCPULoadInfo()
+	if err := processErr(e); err != nil {
+		// exit immediately if error encountered
+		return nil
```
Why are we returning `nil` here when we have actually encountered an error?

wlan0 replied:
I'm returning `nil` to keep up the aesthetics of our cli output. If I returned err, this is how the output looks.

```
ubuntu@ip-172-31-27-137:~/code/go/src/github.com/minio/mc$ ./mc  admin monitor local11111                                                                                            
●  local11111                                                                                                                                                                        
   Server is offline                                                                                                                                                                 
Get http://127.0.0.1:11111/minio/admin/v1/performance?perfType=cpu: dial tcp 127.0.0.1:11111: connect: connection refused 
```

wlan0 replied:
Fixed the formatting to look like this:

```
●  172.17.0.2:9000                                                                                                                                                                   
                                                                                                                                                                                     
   CPU        min        avg      max                                                                                                                                                
   current    0.08%      0.10%    0.10%                                                                                                                                              
   historic   0.02%      0.13%    4.36%                                                                                                                                              
                                                                                                                                                                                     
   MEM        usage                                                                                                                                                                  
   current    71MiB                                                                                                                                                                  
   historic   71MiB                                                                                                                                                                  
                                                                                                                                                                                     
                                                                                                                                                                                     
●  172.17.0.3:9000                                                                                                                                                                   
   Server is offline                                                                                                                                                                 
   Error : Post http://172.17.0.3:9000/minio/peer/remote: dial tcp 172.17.0.3:9000: connect: connection refused                                                                      
                                                                                                                                                                                     
●  172.17.0.4:9000                                                                                                                                                                   
                                                                                                                                                                                     
   CPU        min        avg      max                                                                                                                                                
   current    0.07%      0.08%    0.09%                                                                                                                                              
   historic   0.02%      0.13%    4.38%                                                                                                                                              
                                                                                                                                                                                     
   MEM        usage                                                                                                                                                                  
   current    71MiB                                                                                                                                                                  
   historic   71MiB                                                                                                                                                                  
                                                                                                                                                                                     
                                                                                                                                                                                     
●  172.17.0.5:9000                                                                                                                                                                   
   Server is offline                                                                                                                                                                 
   Error : Post http://172.17.0.5:9000/minio/peer/remote: dial tcp 172.17.0.5:9000: connect: no route to host         
```

krisis commented on cmd/admin-monitor.go:
```diff
@@ -0,0 +1,202 @@
+	memUsages, e := client.ServerMemUsageInfo()
+	if err := processErr(e); err != nil {
+		// exit immediately id error encountered
+		return nil
```
Same question as above.
(nit: s/id/if)

wlan0 replied:
good catch

wlan0 commented on cmd/admin-monitor.go:
```diff
@@ -0,0 +1,202 @@
+
+var adminMonitorCmd = cli.Command{
+	Name:            "monitor",
+	Usage:           "monitor cpu, mem, net and disk statistics",
```
I agree with that. I'm adding the other resources right now. I'll couple the CLI messages
 along with the appropriate PRs.

wlan0 commented on cmd/admin-monitor.go:
```diff
@@ -0,0 +1,202 @@
+	checkAdminMonitorSyntax(ctx)
+
+	// set the console colors
+	console.SetColor("Monitor", color.New(color.FgGreen, color.Bold))
```
sounds good

wlan0 commented on pkg/colorjson/colors.go:
```diff
@@ -0,0 +1,32 @@
+const FgDarkGray = 1 // needs to be followed by color FgBlack to render gray
+
+func init() {
+	console.SetColor("JSONString", color.New(color.FgGreen))
```
Constants here also?

krisis commented on cmd/admin-monitor.go:
```diff
@@ -0,0 +1,252 @@
+}
+
+// serverMonitorMessage holds service status info along with
+// cpu, mem, net and disk monitoristics
```
```suggestion
// cpu and mem statistics
```

wlan0 replied:
OMG. I dont even know how it became monitoristics. Will fix

krisis commented on cmd/admin-monitor.go:
```diff
@@ -0,0 +1,252 @@
+	return nil
+}
+
+type sortWrapper struct {
```
We should order them by their occurrence in the endpoints list.
 I.e in `minio server http://node1/disk1 http://node2/disk2 ...`, `node1` statistics should come before `node2` and so on, left to right.

krisis replied:
Why are we sorting the list of CPU/memory usage of minio servers in ascending order of the server address? What is the rationale? (Note: the corresponding code change is no longer part of this PR)

wlan0 commented:
@krisis @harshavardhana updated based on comments
